### PR TITLE
Milestone database table and functions

### LINKED ISSUE #201: Milestone database table and functions

### PR DESCRIPTION
**Pull request by thibault-martinez**
Fixes #201 and #140

# Test Plan:
Existing and added tests

### CODE REVIEW

File: common/storage/sql/statements.h [hunk 1]
```diff
@@ -51,7 +51,7 @@ retcode_t iota_transactions_select_hashes_approvers_statement(
  * Milestone statements
  */
 
-retcode_t iota_milestone_insert_statement(const milestone_t* milestone,
+retcode_t iota_milestone_insert_statement(const iota_milestone_t* milestone,
```
Comment (tsvisabo) [inline common/storage/sql/statements.h]: a bit annoying, but i think it should be `iota_statement_milestone_insert`

File: common/storage/sql/sqlite3/storage.c [hunk 2]
```diff
@@ -357,3 +371,168 @@ retcode_t iota_stor_transaction_update(const connection_t* const conn,
...
+
+/*
+ * Milestone operations
+ */
+
+retcode_t iota_stor_milestone_store(const connection_t* const conn,
+                                    const milestone_t* data_in) {
+  retcode_t ret = RC_OK;
+  char statement[MILESTONE_MAX_STORE_STATEMENT_SIZE];
+  char const* err_msg = 0;
```
Comment (tsvisabo) [inline common/storage/sql/sqlite3/storage.c]: we should assert `data_in`

File: common/storage/sql/sqlite3/storage.c [hunk 3]
```diff
@@ -357,3 +371,168 @@ retcode_t iota_stor_transaction_update(const connection_t* const conn,
...
+    return RC_SQLITE3_FAILED_FINALIZE;
+  }
+
+  return RC_OK;
+}
+
+retcode_t iota_stor_milestone_exist(const connection_t* const conn,
+                                    const char* index_name,
+                                    const trit_array_p key, bool* exist) {
+  retcode_t ret = RC_OK;
```
Comment (tsvisabo) [inline common/storage/sql/sqlite3/storage.c]: should assert index_name and maybe rename it align with the "load" query

File: common/storage/sql/statements.c [hunk 4]
```diff
@@ -18,6 +18,54 @@
 
 #define SQL_STATEMENTS_ID "sql_statements"
 
+/*
+ * Generic statements
+ */
+
+static retcode_t iota_generic_select_statement(const char *table_name,
```
Comment (tsvisabo) [inline common/storage/sql/statements.c]: rename to iota_statement_generic_select

File: common/storage/sql/statements.c [hunk 5]
```diff
@@ -18,6 +18,54 @@
...
+
+  if (res < 0 || res == statement_cap) {
+    log_error(SQL_STATEMENTS_ID,
+              "Failed in creating statement, statement: %s\n", statement);
+    return RC_SQL_FAILED_WRITE_STATEMENT;
+  }
+  return RC_OK;
+}
+
+static retcode_t iota_generic_exist_statement(const char *table_name,
```
Comment (tsvisabo) [inline common/storage/sql/statements.c]: same

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: common/storage/sql/statements.h (hunk 1 above)
Comment (thibault-martinez) [reply common/storage/sql/statements.h]: 100% agreed

File: common/storage/sql/sqlite3/storage.c (hunk 3 above)
Comment (thibault-martinez) [reply common/storage/sql/sqlite3/storage.c]: `index_name` can actually be `NULL` in an exist statement

Comment (tsvisabo) [reply common/storage/sql/sqlite3/storage.c]: forgot that..

**Review by tsvisabo — APPROVED** [review]

**Pull request merged**